graphql/src/Mu/GraphQL/Quasi.hs

```diff
@@ -0,0 +1,106 @@
+  where
+    scalarToType :: GQL.Name -> Q Type
+    scalarToType (GQL.unName -> "Int")     = [t|'TPrimitive Int32|]
+    scalarToType (GQL.unName -> "Float")   = [t|'TPrimitive Double|]
+    scalarToType (GQL.unName -> "String")  = [t|'TPrimitive T.Text|]
+    scalarToType (GQL.unName -> "Boolean") = [t|'TPrimitive Bool|]
+    scalarToType (GQL.unName -> "ID")      = [t|'TPrimitive UUID|]
+    scalarToType _                         = fail "not well-known scalar types are not supported"
+typeToDec _ (GQL.TypeDefinitionObject objs) = objToDec objs
+  where
+    objToDec :: GQL.ObjectTypeDefinition -> Q Result
+    objToDec (GQL.ObjectTypeDefinition _ nm _ _ flds) =
+      GQLService <$> [t| 'Service $(textToStrLit $ GQL.unName nm) '[]
+          $(typesToList <$> traverse gqlFieldToType flds) |]
+    gqlFieldToType :: GQL.FieldDefinition -> Q Type
+    gqlFieldToType (GQL.FieldDefinition _ (GQL.unName -> fnm) args ftyp _) =
+      [t| 'Method $(textToStrLit fnm) '[]
+            $(typesToList <$> traverse argToType args)
+            'RetSingle $(gtypeToType ftyp) |]
+    argToType :: GQL.InputValueDefinition -> Q Type
+    argToType (GQL.InputValueDefinition _ (GQL.unName -> aname) atype Nothing) =
+      [t| 'ArgSingle ('Just $(textToStrLit aname)) '[] $(gtypeToType atype) |]
+    argToType (GQL.InputValueDefinition _ (GQL.unName -> aname) atype (Just defs)) =
+      [t| 'ArgSingle ('Just $(textToStrLit aname)) '[] {-$(fromGQLValueConst defs)-} $(gtypeToType atype) |]
+typeToDec _ (GQL.TypeDefinitionInterface _)       = fail "interface types are not supported"
+typeToDec _ (GQL.TypeDefinitionUnion _)           = fail "union types are not supported"
+typeToDec _ (GQL.TypeDefinitionEnum enums)        = enumToDecl enums
+  where
+    enumToDecl :: GQL.EnumTypeDefinition -> Q Result
+    enumToDecl (GQL.EnumTypeDefinition _ (GQL.unName -> name) _ symbols) =
+      GQLSchema <$> [t|'DEnum $(textToStrLit name)
+                              $(typesToList <$> traverse gqlChoiceToType symbols)|]
+    gqlChoiceToType :: GQL.EnumValueDefinition -> Q Type
+    gqlChoiceToType (GQL.EnumValueDefinition _ (GQL.unName . GQL.unEnumValue -> c) _) =
+      [t|'ChoiceDef $(textToStrLit c)|]
+typeToDec _ (GQL.TypeDefinitionInputObject inpts) = inputObjToDec inpts
+  where
+    inputObjToDec :: GQL.InputObjectTypeDefinition -> Q Result
+    inputObjToDec (GQL.InputObjectTypeDefinition _ (GQL.unName -> name) _ fields) =
+        GQLSchema <$> [t|'DRecord $(textToStrLit name)
+                                  $(typesToList <$> traverse gqlFieldToType fields)|]
+    gqlFieldToType :: GQL.InputValueDefinition -> Q Type
+    gqlFieldToType (GQL.InputValueDefinition _ (GQL.unName -> fname) ftype _) =
+      [t|'FieldDef $(textToStrLit fname) $(gtypeToType ftype)|]
+
+gtypeToType :: GQL.GType -> Q Type
+gtypeToType (GQL.TypeNamed (GQL.unNullability -> False) (GQL.unName . GQL.unNamedType -> a)) =
```

Comment:
This is not right, since the type could also be a scalar. In that case you need to use `PrimitiveRef`.

Reply:
[This](https://github.com/higherkindness/mu-haskell/blob/master/graphql/exe/Main.hs#L59) is an example of an object returning a primitive instead of an object reference.